src/ar_firewall.erl

```diff
@@ -1,71 +1,144 @@
-%% Should store compiled binary scan objects.
+%% Stores compiled binary scan objects.
 -record(state, {
-	sigs = [] % the set of known signatures to filter against
+	tx_id_sigs = [], % a set of known signatures to filter transaction identifiers against
+	content_sigs = [] % a set of known signatures to filter transaction content against
 }).
 
 %% @doc Start a firewall node.
 start() ->
-	spawn(
-		fun() ->
-			server(#state {
-				sigs = av_sigs:all()
-			})
-		end
-	).
+	Firewall = whereis(default_firewall),
+	case Firewall of
+		undefined ->
+			F = spawn(
+				fun() ->
+					server(#state {
+						tx_id_sigs = av_sigs:load(ar_meta_db:get(transaction_blacklist)),
+						content_sigs = av_sigs:load(ar_meta_db:get(content_policies))
+					})
+				end
+			),
+			erlang:register(default_firewall, F);
+		_ ->
+			do_nothing
+	end.
+
+reload() ->
+	Firewall = whereis(default_firewall),
+	case Firewall of
+		undefined ->
+			do_nothing;
+		_ ->
+			erlang:unregister(default_firewall),
+			exit(Firewall, {shutdown, reloading})
+	end,
+	start().
 
-%% @doc Check that a received TX does not match the firewall rules.
-scan_tx(FwServer, Data) ->
-	ar:report([{scanning_object_of_type, tx}]),
-	FwServer ! {scan_tx, self(), Data},
+%% @doc Check whether a received TX matches the firewall rules.
+scan_tx(TX) ->
+	FwServer = whereis(default_firewall),
+	FwServer ! {scan_tx, self(), TX},
 	receive
 		{scanned_tx, Response} ->
-			ar:report([{scanned_object_of_type, tx}, {response, Response}]),
 			Response
+	after 10000 ->
+		ar:err([{ar_firewall, scan_tx_timeout}, {tx, ar_util:encode(TX#tx.id)}]),
+		accept
 	end.
 
 %% @doc Main firewall server loop.
-%% Receives scan requests and returns whether the given contents matches
+%% Receives scan requests and returns whether the given transaction and its contents match
 %% the set of known 'harmful'/'ignored' signatures.
-server(S = #state { sigs = Sigs } ) ->
+server(S = #state { tx_id_sigs = TXIDSigs, content_sigs = ContentSigs } ) ->
 	receive
-		{scan_tx, Pid, Data} ->
-			Pid ! {scanned_tx, scan_transaction(Data, Sigs)},
+		{scan_tx, Pid, TX} ->
+			Pid ! {scanned_tx, scan_transaction(TX, TXIDSigs, ContentSigs)},
 			server(S)
 	end.
 
-%% @doc Compare a transaction against known bad signatures.
-scan_transaction(TX, Sigs) ->
-	case av_detect:is_infected(TX#tx.data, Sigs) of
-		{true, _} -> reject;
-		false -> accept
+%% @doc Check if a transaction is in the black list and compare its contents against known bad signatures.
+scan_transaction(TX, TXIDSigs, ContentSigs) ->
+	Tags = lists:foldl(
+		fun({K, V}, Acc) ->
+			[{K, ContentSigs},{V, ContentSigs}|Acc]
+		end,
+		[],
+		TX#tx.tags
+	),
+	ScanList = [{ar_util:encode(TX#tx.id), TXIDSigs},{TX#tx.data, ContentSigs},{TX#tx.target, ContentSigs}|Tags],
+	case lists:any(
+		fun({Data, Sigs}) ->
+			case av_detect:is_infected(Data, Sigs) of
+				{true, _} ->
+					ar:info([{ar_firewall, reject_tx}, {tx, ar_util:encode(TX#tx.id)}, {content, Data}]),
```

Comment:
Seems like this will dump the complete binary in the logfile. I assumed error_logger would truncate it, but it doesn't seem to do that.

Reply:
I've rewritten it to log the matched signatures, makes much more sense this way.

Reply:
✅